### PR TITLE
[WIP] Check estimator

### PR DESCRIPTION
**Pull request by jasmainak**
closes https://github.com/glm-tools/pyglmnet/issues/157 and https://github.com/glm-tools/pyglmnet/issues/153

cc @Eric89GXL let's do TDD :)

### CODE REVIEW

**Comment by jasmainak** [comment]
and now I'm stuck ... I have no idea why scikit-learn doesn't like my fixes.

**Comment by Eric89GXL** [comment]
The `sklearn` check looks fragile when dealing with `ndarray`, which seems to be the error source:

https://github.com/scikit-learn/scikit-learn/blob/0.17.X/sklearn/utils/estimator_checks.py#L1414

This is also the check on latest master:

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/estimator_checks.py#L1593

But really you shouldn't have `ndarray` init parameters because they are mutable, which is probably why the `sklearn` folks haven't hit the error before. I assume somewhere you have an `ndarray` default where you should otherwise use tuple...?

File: pyglmnet/pyglmnet.py [hunk 1]
```diff
@@ -138,15 +140,13 @@ class GLM(object):
 
     def __init__(self, distr='poisson', alpha=0.5,
                  Tau=None, group=None,
-                 reg_lambda=None,
+                 reg_lambda=np.logspace(np.log(0.5), np.log(0.01), 10,
+                                        base=np.exp(1)),
```
Comment (Eric89GXL) [inline pyglmnet/pyglmnet.py]: ... and here it is. This is `ndarray`, mutable. Actually turn this into a `tuple` of values (and any other offending ones) and you should be good to go

**Comment by pavanramkumar** [comment]
@jasmainak @Eric89GXL thanks! the discussions in #153 and #157 led to the conclusion that we need two separate classes: 

- `GLM()` which takes in only a single value of `reg_lambda` so that it is compatible with `GridSearchCV()`, and

- `GLMCV()` which can take an ndarray of `reg_lambda` which is used for warm restarts.

This scheme preserves scikit-learn's conventions in `ElasticNet()` and `ElasticNetCV()`. Let's implement at the sprint?

**Comment by pavanramkumar** [comment]
@jasmainak still haven't looked into the details, but I think we don't need to interpret scikit-learn compatibility literally. we need to make sure that people can use relatively common things that they may use with estimators such as `GridSearchCV()` and maybe `Pipeline()`, even if it doesn't pass formal estimator checks.

Passing all  formal checks is a rabbit hole we don't want to get stuck in :)

**Comment by jasmainak** [comment]
Another idea might be to have `scikit-learn` as an optional dependency. A construct like this might do:

```python
has_sklearn = False
try:
    import sklearn

if has_sklearn:
   from sklearn.utils.validation import check_array
   check_array(X)
```

Basically, if the user is using stuff from scikit-learn for cross-validation, pipeline etc., then we can assume they have scikit-learn installed. In that case, it is okay to use the utilities that `scikit-learn` already provides to pass their checks. It will be an optional / soft dependency in that case.

**Pull request closed without merging**